provider-middleware/daily_program_history.go

```diff
@@ -0,0 +1,189 @@
+package main
+
+import (
+	"UnlockEdv2/src/models"
+	"context"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+func InsertDailyProgHistory(ctx context.Context, db *gorm.DB) error {
+	if err := InsertDailyProgramsFacilitiesHistory(ctx, db); err != nil {
+		log.Errorln("error inserting daily programs facilities history")
+		return err
+	}
+	if err := InsertDailyProgramFacilitiesHistory(ctx, db); err != nil {
+		log.Errorln("error inserting daily program facilities history")
+		return err
+	}
+	if err := InsertDailyProgramFacilityHistory(ctx, db); err != nil {
+		log.Errorln("error inserting daily program facility history")
+		return err
+	}
+	return nil
+}
+
+func InsertDailyProgramsFacilitiesHistory(ctx context.Context, db *gorm.DB) error {
+	var history models.DailyProgramsFacilitiesHistory
+	var programStats struct {
+		TotalPrograms         int64 `json:"total_programs"`
+		TotalActivePrograms   int64 `json:"total_active_programs"`
+		TotalArchivedPrograms int64 `json:"total_archived_programs"`
+	}
+	if err := db.WithContext(ctx).Model(&models.Program{}).
+		Select(`
+			COUNT(*) AS total_programs,
+			SUM(CASE WHEN is_active = true THEN 1 ELSE 0 END) AS total_active_programs,
+			SUM(CASE WHEN archived_at IS NOT NULL THEN 1 ELSE 0 END) AS total_archived_programs
+		`).
+		Scan(&programStats).Error; err != nil {
+		log.Errorln("error getting total, active, archived programs")
+		return err
+	}
+	history.TotalPrograms = programStats.TotalPrograms
+	history.TotalActivePrograms = programStats.TotalActivePrograms
+	history.TotalArchivedPrograms = programStats.TotalArchivedPrograms
+
+	var enrollmentStats struct {
+		TotalEnrollments int64 `json:"total_enrollments"`
+		TotalCompletions int64 `json:"total_completions"`
+	}
+	if err := db.WithContext(ctx).Model(&models.ProgramClassEnrollment{}).
+		Select(`
+			COUNT(*) AS total_enrollments,
+			SUM(CASE WHEN enrollment_status = 'Completed' THEN 1 ELSE 0 END) AS total_completions
+		`).
+		Scan(&enrollmentStats).Error; err != nil {
+		log.Errorln("error getting total, completed enrollments")
+		return err
+	}
+	history.TotalEnrollments = enrollmentStats.TotalEnrollments
+	history.TotalCompletions = enrollmentStats.TotalCompletions
+
+	var totalProgramOfferings int64
+	if err := db.WithContext(ctx).Model(&models.FacilitiesPrograms{}).
+		Select("COUNT(*) AS total_program_offerings").
+		Joins("JOIN programs ON programs.id = facilities_programs.program_id AND programs.is_active = true").
```

Comment:
@calisio Do we need to filter out programs where the archived_at IS NOT NULL so that we don't count  archived ones?